src/CPUSolver.cpp

```diff
@@ -38,132 +38,52 @@ int CPUSolver::getNumThreads() {
-               fsr_id, energy_group);
-
-  if (energy_group-1 >= _num_groups)
-    log_printf(ERROR, "Unable to return a source for FSR ID = %d "
-               "in energy group %d since the solver only has %d energy "
-               "groups", fsr_id, energy_group, _num_groups);
-
-  if (energy_group <= 0)
-    log_printf(ERROR, "Unable to return a source for FSR ID = %d "
-               "in energy group %d since energy groups are greater than "
-               "or equal to 1", fsr_id, energy_group);
-
-  Material* material = _FSR_materials[fsr_id];
-  FP_PRECISION* nu_sigma_f = material->getNuSigmaF();
-  FP_PRECISION* chi = material->getChi();
-  FP_PRECISION fission_source = 0.0;
-  FP_PRECISION scatter_source = 0.0;
-  FP_PRECISION total_source = 0.0;
-
-  /* Compute fission source for each group */
-  if (material->isFissionable()) {
-    for (int e=0; e < _num_groups; e++)
-      fission_source += _scalar_flux(fsr_id,e) * nu_sigma_f[e];
-
-    fission_source /= _k_eff;
-  }
-
-  for (int g=0; g < _num_groups; g++)
-    scatter_source += material->getSigmaSByGroupInline(g,energy_group-1)
-        * _scalar_flux(fsr_id,g);
-
-  /* Compute the total source */
-  total_source = (fission_source * chi[energy_group-1] + scatter_source) *
-      ONE_OVER_FOUR_PI;
-
-  return total_source;
-}
-
-
-/**
- * @brief Return a scalar flux array indexed by FSR IDs and energy groups.
- * @details This energy groups are the innermost index, while the FSR ID is
- *         the outermost index.
- * @return an array of flat source region scalar fluxes
+ * @brief Sets the number of shared memory OpenMP threads to use (>0).
+ * @param num_threads the number of threads
  */
-FP_PRECISION* CPUSolver::getFSRScalarFluxes() {
+void CPUSolver::setNumThreads(int num_threads) {
 
-  if (_scalar_flux == NULL)
-    log_printf(ERROR, "Unable to returns the Solver's FSR scalar flux array "
-               "since it has not yet been allocated in memory");
+  if (num_threads <= 0)
+    log_printf(ERROR, "Unable to set the number of threads to %d "
+               "since it is less than or equal to 0", num_threads);
 
-  return _scalar_flux;
+  /* Set the number of threads for OpenMP */
+  _num_threads = num_threads;
+  omp_set_num_threads(_num_threads);
 }
 
 
 /**
- * @brief Sets the number of shared memory OpenMP threads to use (>0).
- * @param num_threads the number of threads
+ * @brief Assign a fixed source for a flat source region and energy group.
+ * @details Fixed sources should be scaled to reflect the fact that OpenMOC 
+ *          normalizes the scalar flux such that the total energy- and 
+ *          volume-integrated production rate sums to 1.0.
+ * @param fsr_id the flat source region ID
+ * @param group the energy group
+ * @param source the volume-averaged source in this group
  */
-void CPUSolver::setNumThreads(int num_threads) {
+void CPUSolver::setFixedSourceByFSR(int fsr_id, int group, 
```

Comment:
Ok, that is a good point, I hadn't considered that possibility. So you think that simply over-riding the source in the FSR and printing a ``WARNING`` message is sufficient? I just added the message. If you can think of a good use case which would make it prohibitively difficult to assign sources in this way, we could add an additional method parameter ``increment=False`` which would allow the user to increment any source already assigned to the FSR if they wish.